app/controllers/application_controller/timelines.rb

```diff
@@ -102,42 +102,54 @@ def tl_build_timeline_report_options
       end
 
       temp_clause = @tl_record.event_where_clause(@tl_options.evt_type)
+      conditions = [temp_clause[0], "timestamp >= ?", "timestamp <= ?"]
+      parameters = temp_clause[1..-1] + [from_dt, to_dt]
 
-      cond = "( "
-      cond = cond << temp_clause[0]
-      params = temp_clause.slice(1, temp_clause.length)
-
-      event_set = @tl_options.event_set
-      if !event_set.empty?
-        if @tl_options.policy_events? && @tl_options.policy.result != "both"
-          where_clause = [") and (timestamp >= ? and timestamp <= ?) and (event_type in (?)) and (result = ?)",
-                          from_dt,
-                          to_dt,
-                          event_set.flatten,
-                          @tl_options.policy.result]
-        else
-          where_clause = [") and (timestamp >= ? and timestamp <= ?) and (event_type in (?))",
-                          from_dt,
-                          to_dt,
-                          event_set.flatten]
-        end
-      else
-        where_clause = [") and (timestamp >= ? and timestamp <= ?)",
-                        from_dt,
-                        to_dt]
-      end
-      cond << where_clause[0]
+      tl_add_event_type_conditions(conditions, parameters)
+      tl_add_policy_conditions(conditions, parameters) if @tl_options.policy_events?
 
-      params2 = where_clause.slice(1, where_clause.length - 1)
-      params = params.concat(params2)
-      @report.where_clause = [cond, *params]
+      condition = "(" + conditions.reduce { |a, c| "#{a}) and (#{c}" } + ")"
+      @report.where_clause = [condition] + parameters
       @report.rpt_options ||= {}
-      @report.rpt_options[:categories] =
-        @tl_options.management_events? ? @tl_options.management.categories : @tl_options.policy.categories
+      @report.rpt_options[:categories] = @tl_options.categories
       @title = @report.title
     end
   end
 
+  def tl_add_event_type_conditions(conditions, parameters)
+    tl_add_event_type_inclusions(conditions, parameters)
+
+    excludes = @tl_options.get_set(:exclude_set)
+    unless excludes.empty?
+      conditions << "event_type not in (?)"
+      parameters << excludes
+    end
+  end
+
+  def tl_add_event_type_inclusions(conditions, parameters)
+    expressions = []
+    @tl_options.get_set(:regexes).each do |regex|
+      expressions << "event_type ~ ?"
```

Comment:
This assumes that all Ruby regexps can be SQL regexps, which I don't believe is true.

Reply:
PostgreSQL's support is quite powerful. I will compare the capabilities of the Ruby's vs. PostgreSQL's regular expressions and report back. If nothing else, we can document the limitations somewhere and inform settings authors what can and what cannot be used.

Reply:
Where should I put the specs: in the core repo where the `EventStream` lives or in the UI repo?

As for the regexp modifiers, Ruby currently [supports four][options], but the only one that would be of any use in our case is `i` that controls case sensitivity. And adding support for this option is quite easy, we simply need to use proper PostgreSQL operator as you already stated.

   [options]: https://ruby-doc.org/core-2.3.1/Regexp.html#class-Regexp-label-Options

Reply:
I think the "check the event_groups spec" should be in the core repo.  Ultimately it should be in the provider repos so that provider authors get a failure when they add it, but I'm not sure how to do that.  Maybe we can have the spec in core, but run it in the provider repos somehow?